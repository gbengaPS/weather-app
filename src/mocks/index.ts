export const citiesMock = {
  base: "stations",
  clouds: {all: 0},
  cod: 200,
  coord: {lon: 116.4, lat: 39.91},
  dt: 1602226045,
  id: 1816670,
  main: {temp: 23.52, feels_like: 22.22, temp_min: 23.33, temp_max: 23.89, pressure: 1014, humidity: 47},
  name: "Beijing",
  sys: {type: 3, id: 2000022, country: 'CN', sunrise: 1602195502, sunset: 1602236679},
  timezone: 28800,
  visibility: 10000,
  weather: [{id: 800, main: "Clear", description: "clear sky", icon: "01d"}],
  wind: {speed: 2.55, deg: 209},
};

export const initialStateMock =  {
  favorites: { 'Beijing': { data: citiesMock } },
  weatherInfo: [],
  notes: { 'Beijing': 'hello world' },
  isLoading: false
}