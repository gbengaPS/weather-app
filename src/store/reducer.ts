const reducer = (state: any, action: any) => {
  return {}
}

export default reducer;