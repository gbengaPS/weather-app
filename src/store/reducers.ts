import { REMOVE_WEATHER_INFO, ADD_WEATHER_INFO, ADD_FAVORITE, REMOVE_FAVORITE } from '../constants';
import { FavoritesMap, WeatherInfo } from '../types';
import { saveFavorites } from '../helpers';

export const weatherInfoReducer = (state: WeatherInfo[], action: any) => {
  switch(action.type) {
    case REMOVE_WEATHER_INFO:
      const filteredInfo = state.filter((info: WeatherInfo) => {
        return info.data.name !== action.cityName;
      });
      localStorage.setItem('largestCitiesInfo', JSON.stringify(filteredInfo))
      return filteredInfo;

    case ADD_WEATHER_INFO:
      return [ ...state, ...action.weatherInfo ];
    default:
      return state;
  }
}

export const favoritesReducer = (state: FavoritesMap, action: any) => {
  switch(action.type) {
    case ADD_FAVORITE:
      const newState = { ...state, ...action.favoriteItem };
      saveFavorites(newState)
      return newState;

    case REMOVE_FAVORITE:
      const stateCopy = {...state};
      delete stateCopy[action.cityName];
      saveFavorites(stateCopy);
      return stateCopy;
    default:
      return state
  }
}