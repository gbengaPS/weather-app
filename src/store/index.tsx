import React, { createContext, useReducer } from 'react';
import { FavoritesMap, WeatherInfo } from '../types';
import { weatherInfoReducer, favoritesReducer } from './reducers';

interface InitialStateType {
	favorites: FavoritesMap
	weatherInfo: WeatherInfo[]
}

const initialState  = {
	favorites: {},
	weatherInfo: []
}

const AppContext = createContext<{
    state: InitialStateType;
    dispatch: React.Dispatch<any>;
  }>({
    state: initialState,
    dispatch: () => null
  });

const mainReducer = (state: InitialStateType, action: any) => ({
  weatherInfo: weatherInfoReducer(state.weatherInfo, action),
  favorites: favoritesReducer(state.favorites, action)
});

const AppProvider: React.FC = ({ children }) => {
  const [state, dispatch] = useReducer(mainReducer, initialState);

  return (
      <AppContext.Provider value={{state, dispatch}}>
      {children}
      </AppContext.Provider>
  )
}
  
  export { AppContext, AppProvider };