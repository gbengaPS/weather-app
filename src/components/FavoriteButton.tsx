import React, { useContext } from 'react';
import { AppContext } from '../store';
import {
  ADD_FAVORITE,
  REMOVE_FAVORITE,
  ADD_WEATHER_INFO,
  REMOVE_WEATHER_INFO,
  largestCities
} from '../constants';
import { CityDetails } from '../types';


interface FavoriteProps {
  cityDetails: CityDetails 
}

const FavoriteButton = (props: FavoriteProps) => {
  const { cityDetails } = props;
  const { state, dispatch } = useContext(AppContext);

  let iconName = 'fa-heart-o';
  let color = '#000';
  
  if (state.favorites[cityDetails.name]) {
    iconName = 'fa-heart';
    color = '#ec6e4d';
  }

  const hanldeClick = () => {
    const cityName = cityDetails.name;
    const isOneOfLargestCities = largestCities.find((name) => name === cityName);
  
    if(state.favorites[cityName]) {
      dispatch({ type: REMOVE_FAVORITE, cityName });
      isOneOfLargestCities ?
        dispatch({ type: ADD_WEATHER_INFO, weatherInfo: [{ data: cityDetails }] }) :
        dispatch({ type: REMOVE_WEATHER_INFO, cityName });
    } else {
      const favoriteItem = { [cityName]: { data: cityDetails } };
      dispatch({ type: ADD_FAVORITE, favoriteItem });
      dispatch({ type: REMOVE_WEATHER_INFO, cityName });
    }
  }
  
  return (
    <button
      className="app-button"
      onClick={hanldeClick}
      style={{color}}
    >
      <i className={`fa ${iconName}`} aria-hidden="true"></i>
    </button>
  )

};


export default FavoriteButton