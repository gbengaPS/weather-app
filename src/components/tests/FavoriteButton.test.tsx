import React from 'react';
import { render, cleanup, fireEvent } from '@testing-library/react';
import FavoriteButton from '../FavoriteButton';
import { AppContext } from '../../store';
import { citiesMock, initialStateMock as state } from '../../mocks';


describe('FavoriteButton', () => {
  const props = { cityDetails: citiesMock };

  const dispatch = jest.fn();

  afterEach(cleanup);

  it('should match snapshot', () => {
    const { container } = render(<FavoriteButton {...props}/>);
    expect(container).toMatchSnapshot();
  });

  it('should match snapshot when cityName is a favorite', () => {

    const { container } = render(
      <AppContext.Provider value={{dispatch, state }}>
        <FavoriteButton {...props}/>
      </AppContext.Provider>
    );

    expect(container).toMatchSnapshot();
  });

  it('should call dispatch when favorite button is clicked', () => {
    const { getByTestId } = render(
      <AppContext.Provider value={{dispatch, state }}>
        <FavoriteButton {...props}/>
      </AppContext.Provider>
    );
    
    fireEvent.click(getByTestId('favorite-button'));
    expect(dispatch).toHaveBeenCalledTimes(2);
  });

  it('should call dispatch when city is not in favorite list and button is clicked', () => {
    props.cityDetails.name = 'Lagos';
    const { getByTestId } = render(
      <AppContext.Provider value={{dispatch, state }}>
        <FavoriteButton {...props}/>
      </AppContext.Provider>
    );
    
    fireEvent.click(getByTestId('favorite-button'));
    expect(dispatch).toHaveBeenCalled();
  });
});