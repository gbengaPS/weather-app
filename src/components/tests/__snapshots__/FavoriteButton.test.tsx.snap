// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FavoriteButton should match snapshot 1`] = `
<div>
  <button
    class="app-button"
    style="color: rgb(0, 0, 0);"
  >
    <i
      aria-hidden="true"
      class="fa fa-heart-o"
      data-testid="favorite-button"
    />
  </button>
</div>
`;

exports[`FavoriteButton should match snapshot when cityName is a favorite 1`] = `
<div>
  <button
    class="app-button"
    style="color: rgb(236, 110, 77);"
  >
    <i
      aria-hidden="true"
      class="fa fa-heart"
      data-testid="favorite-button"
    />
  </button>
</div>
`;
